polkadot/runtime/parachains/src/assigner_bulk/mod.rs

```diff
@@ -241,37 +249,43 @@ impl<T: Config> AssignmentProvider<BlockNumberFor<T>> for Pallet<T> {
 
 		Workload::<T>::mutate(core_idx, |core_state| {
 			Self::ensure_workload(now, core_idx, core_state);
-
-			core_state.pos = core_state.pos % core_state.assignments.len();
-			let (a_type, a_state) = &mut core_state
-				.assignments
-				.get(core_state.pos)
-				.expect("We limited pos to the size of the vec one line above. qed");
-
-			// advance:
-			a_state.remaining -= core_state.step;
-			if a_state.remaining < step {
-				// Assignment exhausted, need to move to the next and credit remaining for next
-				// round.
-				core_state.pos += 1;
-				// Reset to ratio + still remaining "credits":
-				a_state.remaining += a_state.ratio;
-			}
-
-			match a_type {
-				CoreAssignment::Idle => return None,
-				CoreAssignment::Pool =>
-					return <assigner_on_demand::Pallet<T> as AssignmentProvider<
-						BlockNumberFor<T>,
-					>>::pop_assignment_for_core(core_idx),
-				CoreAssignment::Task(para_id) => return Some(BulkAssignment::Bulk(para_id)),
+			match core_state {
```

Comment:
We can avoid all of this code being nested by short circuiting the function on `None`.

Reply:
Actually the correct solution is, that this should be a `ValueQuery` not an `OptionQuery`.